vimiv/imutils/exif.py

```diff
@@ -10,139 +10,361 @@
+    """Implementation of ExifHandler based on depreciated piexif."""
+
+    def __init__(self, filename=""):
+        self._metadata = None
+
+        try:
+            self._metadata = piexif.load(filename)
+        except FileNotFoundError:
+            _logger.debug("File %s not found", filename)
+            return
 
-@check_piexif("")
-def exif_date_time(filename) -> str:
-    """Exif creation date and time of filename."""
-    with contextlib.suppress(piexif.InvalidImageDataError, FileNotFoundError, KeyError):
-        exif_dict = piexif.load(filename)
-        return exif_dict["0th"][piexif.ImageIFD.DateTime].decode()
-    return ""
+    def get_formatted_exif(self) -> Dict[str, str]:
 
+        desired_keys = [
+            e.strip() for e in api.settings.metadata.current_keyset.value.split(",")
+        ]
+        _logger.debug(f"Read metadata.current_keys {desired_keys}")
 
-class ExifInformation(dict):
-    """Dictionary to load and store exif information of a single image.
+        exif = dict()
 
-    Exif information is loaded and formatted upon construction. Afterwards the class
-    behaves like a regular dictionary.
+        try:
 
-    Attributes:
-        _exif: Loaded piexif exif information dictionary if any.
+            for ifd in self._metadata:
+                if ifd == "thumbnail":
+                    continue
+
+                for tag in self._metadata[ifd]:
+                    keyname = piexif.TAGS[ifd][tag]["name"]
+                    keytype = piexif.TAGS[ifd][tag]["type"]
+                    val = self._metadata[ifd][tag]
+                    _logger.debug(
+                        f"name: {keyname}\
+                        type: {keytype}\
+                        value: {val}\
+                        tag: {tag}"
+                    )
+                    if keyname not in desired_keys:
+                        _logger.debug(f"Ignoring key {keyname}")
+                        continue
+                    if keytype in (
+                        piexif.TYPES.Byte,
+                        piexif.TYPES.Short,
+                        piexif.TYPES.Long,
+                        piexif.TYPES.SByte,
+                        piexif.TYPES.SShort,
+                        piexif.TYPES.SLong,
+                        piexif.TYPES.Float,
+                        piexif.TYPES.DFloat,
+                    ):  # integer and float
+                        exif[keyname] = (keyname, val)
+                    elif keytype in (
+                        piexif.TYPES.Ascii,
+                        piexif.TYPES.Undefined,
+                    ):  # byte encoded
+                        exif[keyname] = (keyname, val.decode())
+                    elif keytype in (
+                        piexif.TYPES.Rational,
+                        piexif.TYPES.SRational,
+                    ):  # (int, int) <=> numerator, denominator
+                        exif[keyname] = (keyname, f"{val[0]}/{val[1]}")
```

Comment:
Is there any reason why we switched to a dictionary of the form `key -> keyname, value` later in the py3exiv2 version? I don't see us using the actual key anywhere in the formatted metadata, so the initial `keyname -> value` mapping would suffice, no? This would simplify the dict as well as fix the type annotation :smile:

Reply:
Yes, there is a reason why I index the dict this way. `keyname` is not unique among the metadata (e.g. there is `Iptc.Application2.Copyright` and `Exif.Image.Copyright`, so `Copyright` is not suitable as a key).

What we could instead of using `key -> keyname, value`, we could use a `key -> value` dict and extract the `key_name` from the `key` in `vimiv/metadatawidget`. This should work because `key` has the format `familyKey/groupKey/keyName`.

Reply:
Ah, yes, this makes a lot of sense. Thanks for the explanation!

In this case I would add a new type for the dictionary at the top of the module, i.e.
```
_logger = log.module_logger(__name__)

ExifDictT = Dict[Any, Tuple[str, str]]
```
and use this for the return value type annotation of `get_formatted_exif` instead of `ExifDictT`. For completeness we could also ensure that the value is always of type string. This is currently not the case for integer and float in the piexif case and possibly not for the `raw_value` in the py3exiv2 case, this I am not sure about. I suppose a simple `str(...)` should suffice.

Reply:
Thanks for pointing this out concerning the strings. I have fixed that as well as the dict type in e2ebcdf32eec9a14763e555a212c3bdccfd618c6